src/urls/package-url-handler.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {lookupDependencyMapping} from '../manifest-converter';
+
+import {ConvertedDocumentFilePath, ConvertedDocumentUrl, OriginalDocumentUrl, PackageType} from './types';
+import {UrlHandlerInterface} from './url-handler-interface';
+import {getRelativeUrl} from './util';
+
+
+/**
+ * Handle URLs in a single "package-based" layout. This converter should be used
+ * to convert a single package, where all existing Bower dependencies are
+ * installed in a "bower_components/" sub-directory inside the main package
+ * directory.
+ */
+export class PackageUrlHandler implements UrlHandlerInterface {
+  readonly packageName: string;
+  readonly packageType: PackageType;
+
+  /**
+   * Helper function to check if a file URL is internal to the main package
+   * being converted (vs. a dependency).
+   */
+  static isUrlInternalToPackage(url: ConvertedDocumentUrl|OriginalDocumentUrl|
+                                ConvertedDocumentFilePath) {
+    return !url.startsWith('bower_components/') &&
+        !url.startsWith('./node_modules/');
+  }
+
+  constructor(packageName: string, packageType?: PackageType) {
+    this.packageName = packageName;
+    this.packageType = packageType || 'element';
+  }
+
+  /**
+   * Get the converted NPM name for a package, given the original document URL
+   * of any file that lives in that package.
+   */
+  getPackageNameForUrl(url: OriginalDocumentUrl) {
+    // If url contains no directories it must be internal. Explicitly check
+    // `basePackageName` here since it is needed below.
+    const urlParts = url.split('/');
+    const basePackageName = urlParts[1] as (string | undefined);
+    if (!basePackageName || PackageUrlHandler.isUrlInternalToPackage(url)) {
+      return this.packageName;
+    }
+    // For a a single package layout, the Bower package name is included in
+    // the URL. ie: bower_components/PACKAGE_NAME/...
+    // Check the dependency map to get the new NPM name for the package.
+    const depInfo = lookupDependencyMapping(basePackageName);
+    if (!depInfo) {
+      return basePackageName;
+    }
+    return depInfo.npm;
+  }
+
+  /**
+   * Get the "type" for the package where a file lives, based on it's URL.
+   */
+  getPackageTypeForUrl(url: OriginalDocumentUrl) {
```

Comment:
This isn't really a package type so much as a URL type. When is this method called? If it's per-package rather than per URL, could we get rid of this concept at the package level?

Reply:
Yes, this is planned but v. v. much out of scope for this PR.